The final section of this assignment requires you to modify the distance-weighted k-NN algorithm, developed in Part 2, so that it can be used for a regression problem.

In [221]:
import numpy as np
import math

In [222]:
train_dataset  = np.genfromtxt('trainingData_regression.csv', delimiter = ',')
train_data  = np.delete(train_dataset,12,axis = 1)
train_class = train_dataset[:,12]
test_dataset  = np.genfromtxt('testData_regression.csv', delimiter = ',')
test_data  = np.delete(test_dataset,12,axis = 1)
test_class = test_dataset[:,12]

In [223]:
def calculateDistances(query_instance, feature_list):
    feature_difference = feature_list - query_instance
    euclidena_distance = np.sqrt(np.sum(np.square(feature_difference),axis=1))
    sorted_distance_index = np.argsort(euclidena_distance)
    return euclidena_distance, sorted_distance_index

In [273]:
k = 3

In [274]:
results = np.apply_along_axis(calculateDistances,1, test_data,train_data)

In [275]:
sorted_indicies = results[:,1].astype('int32') 
distances = results[:,0]

In [300]:
sorted_indicies

array([[4364, 4260,   18, ...,  232, 6252, 1948],
       [3940, 6068, 4536, ..., 3496, 2182, 2086],
       [2422, 6271, 2313, ..., 2006,  334, 2541],
       ...,
       [5060, 3636, 6142, ..., 5852, 3155, 1948],
       [5301, 2978, 3376, ...,   92,  340,   11],
       [6021, 6039, 1991, ..., 1173, 6074, 4270]])

In [320]:
# Contains all the indicies representing the minimum euclidena distance
minimum_dist = sorted_indicies[:,:k]

In [322]:
minimum_dist

array([[4364, 4260,   18],
       [3940, 6068, 4536],
       [2422, 6271, 2313],
       ...,
       [5060, 3636, 6142],
       [5301, 2978, 3376],
       [6021, 6039, 1991]])

In [278]:
test_dist = distances.copy()

In [279]:
test_dist.sort(axis=1)

In [280]:
my_dist = test_dist[:,:k]  

In [281]:
# Thes are the K nigbour distances from the query instance
my_dist

array([[1.45810197, 1.55958863, 1.64685987],
       [2.26900575, 2.44561223, 2.46743592],
       [1.75701119, 1.94464648, 2.13323388],
       ...,
       [1.60218554, 1.7943869 , 2.08417693],
       [2.47225923, 2.49284388, 2.546252  ],
       [1.69731469, 1.75485118, 1.89387586]])

In [282]:
# Numpy array to store the classes predicted for the test data
prediction = train_class[minimum_dist]

my_dist =  The distances of the K neighbours from the query instance
prediction = The actual regression values of the k neighbours

Hence<br>
For regression value -149.76192018 the distance is 1.45810197<br>
For regression value -214.1207185 the distance is 1.55958863 

and so on

In [283]:
n = 3

In [284]:
final_result_prediction = np.divide(np.sum(np.multiply(1/pow(my_dist,n),prediction),axis=1),np.sum(1/pow(my_dist,n),axis=1))

In [285]:
final_result_prediction

array([-163.1038996 ,   44.83394934,   -6.36912554, ...,  148.68068328,
       -132.10867755,  216.93290105])

### R square matrix calculation

#### Sum of squared residuals

here we need to subtract prediction value from the actual value , take its square and sum all of the squares

In [286]:
# Predicted resultls
final_result_prediction

array([-163.1038996 ,   44.83394934,   -6.36912554, ...,  148.68068328,
       -132.10867755,  216.93290105])

In [287]:
# Actual Results
test_class

array([-174.08350009,   17.81557908,  -20.08799485, ...,  112.4585218 ,
       -185.80280705,  240.30808959])

In [288]:
# Sum of squared residuals. The numerator
SSR = np.sum(np.square(np.subtract(final_result_prediction,test_class)))

In [289]:
SSR

8067648.406305749

#### Calculation for total sum of squares

here we will subtract the average value of actual values from the actual values and square them and then add all

In [290]:
test_class


array([-174.08350009,   17.81557908,  -20.08799485, ...,  112.4585218 ,
       -185.80280705,  240.30808959])

In [291]:
average = np.average(test_class)
average

0.564594108802638

In [292]:
SST = np.sum(np.square(np.subtract(test_class,np.average(test_class))))

In [293]:
SST

44828711.24895885

In [294]:
SSR/SST

0.1799661016686738

In [295]:
(1-(SSR/SST))*100

82.00338983313262

### Practice

In [296]:
prediction[0]

array([-149.76192018, -214.12071851, -122.25775817])

In [297]:
my_dist[0]

array([1.45810197, 1.55958863, 1.64685987])

In [63]:
pow(my_dist[0],2)

array([2.12606136, 2.43231669, 2.71214744])

In [65]:
1/pow(my_dist[0],2)

array([0.47035331, 0.41113067, 0.36871152])

In [66]:
1/pow(my_dist[0],2)*prediction[0]

array([-70.44101502, -88.03159526, -45.07784369])

In [69]:
result = 1/pow(my_dist,2)*prediction

In [73]:
result

array([[-70.44101502, -88.03159526, -45.07784369],
       [ -0.81879819,  26.05941777,  -1.00468786],
       [ -8.97241325,  -3.05610278,   8.49404975],
       ...,
       [ 63.45273281,  50.25842798,  22.29022024],
       [-27.0587465 , -28.478577  ,  -7.39483287],
       [ 57.65806844,  96.81726727,  51.70508468]])

In [80]:
numerator = np.sum(result,axis =1)
numerator

array([-203.55045397,   24.23593172,   -3.53446629, ...,  136.00138104,
        -62.93215637,  206.1804204 ])

In [82]:
pow(my_dist[0],2)

array([2.12606136, 2.43231669, 2.71214744])

In [83]:
1/pow(my_dist[0],2)

array([0.47035331, 0.41113067, 0.36871152])

In [87]:
a = np.sum(1/pow(my_dist[0],2)*prediction[0])
a

-203.5504539721485

In [94]:
b = np.sum(1/pow(my_dist[0],2))
b

1.250195503479115

In [90]:
a/b

-162.81489847443598

In [138]:
v = np.sum(np.multiply(1/pow(my_dist,2),prediction),axis=1)/np.sum(1/pow(my_dist,2),axis=1)

In [139]:
v

array([-162.81489847,   46.10376619,   -4.37373469, ...,  146.18310844,
       -131.44526438,  216.88431953])

In [136]:
a = np.sum(np.multiply(1/pow(my_dist,2),prediction),axis=1)
a

array([-203.55045397,   24.23593172,   -3.53446629, ...,  136.00138104,
        -62.93215637,  206.1804204 ])

In [137]:
b= np.sum(1/pow(my_dist,2),axis=1)
b

array([1.2501955 , 0.52568225, 0.80811173, ..., 0.93034949, 0.47877082,
       0.95064697])

In [140]:
a/b

array([-162.81489847,   46.10376619,   -4.37373469, ...,  146.18310844,
       -131.44526438,  216.88431953])

In [107]:
a = 1/pow(my_dist,2)

In [108]:
a

array([[0.47035331, 0.41113067, 0.36871152],
       [0.19423559, 0.16719559, 0.16425108],
       [0.32392983, 0.26443484, 0.21974705],
       ...,
       [0.38956002, 0.31057595, 0.23021352],
       [0.16361081, 0.16091993, 0.15424008],
       [0.3471165 , 0.32472776, 0.27880271]])

In [109]:
cc = np.sum(1/pow(my_dist,2),axis=1)
cc

array([1.2501955 , 0.52568225, 0.80811173, ..., 0.93034949, 0.47877082,
       0.95064697])